File: rust/fory-core/src/serializer/enum_.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::ensure;
use crate::error::Error;
use crate::meta::FieldInfo;
use crate::resolver::context::{ReadContext, WriteContext};
use crate::serializer::{ForyDefault, Serializer};
use crate::types::{RefFlag, RefMode, TypeId};
use crate::TypeResolver;

#[inline(always)]
pub fn actual_type_id(type_id: u32, register_by_name: bool, _compatible: bool) -> u32 {
    if register_by_name {
        TypeId::NAMED_ENUM as u32
    } else {
        (type_id << 8) + TypeId::ENUM as u32
    }
}

#[inline(always)]
pub fn write<T: Serializer>(
    this: &T,
    context: &mut WriteContext,
    ref_mode: RefMode,
    write_type_info: bool,
) -> Result<(), Error> {
    if ref_mode != RefMode::None {
        context.writer.write_i8(RefFlag::NotNullValue as i8);
    }
    if write_type_info {
        T::fory_write_type_info(context)?;
    }
    this.fory_write_data(context)
}

#[inline(always)]
pub fn write_type_info<T: Serializer>(context: &mut WriteContext) -> Result<(), Error> {
    let type_id = T::fory_get_type_id(context.get_type_resolver())?;
    context.writer.write_varuint32(type_id);
    let is_named_enum = type_id & 0xff == TypeId::NAMED_ENUM as u32;
    if !is_named_enum {
        return Ok(());
    }
    let rs_type_id = std::any::TypeId::of::<T>();
    if context.is_share_meta() {
        // Write type meta inline using streaming protocol
        context.write_type_meta(rs_type_id)?;
    } else {
        let type_info = context.get_type_resolver().get_type_info(&rs_type_id)?;
        let namespace = type_info.get_namespace();
        let type_name = type_info.get_type_name();
        context.write_meta_string_bytes(namespace)?;
        context.write_meta_string_bytes(type_name)?;
    }
    Ok(())
}

#[inline(always)]
pub fn read<T: Serializer + ForyDefault>(
    context: &mut ReadContext,
    ref_mode: RefMode,
    read_type_info: bool,
) -> Result<T, Error> {
    let ref_flag = if ref_mode != RefMode::None {
        context.reader.read_i8()?
    } else {
        RefFlag::NotNullValue as i8
    };
    if ref_flag == RefFlag::Null as i8 {
        Ok(T::fory_default())
    } else if ref_flag == (RefFlag::NotNullValue as i8) || ref_flag == (RefFlag::RefValue as i8) {
        if read_type_info {
            T::fory_read_type_info(context)?;
        }
        T::fory_read_data(context)
    } else if ref_flag == (RefFlag::Ref as i8) {
        Err(Error::invalid_ref("Invalid ref, enum type is not a ref"))
    } else {
        Err(Error::invalid_data(format!(
            "Unknown ref flag: {}",
            ref_flag
        )))
    }
}

#[inline(always)]
pub fn read_type_info<T: Serializer>(context: &mut ReadContext) -> Result<(), Error> {
    let local_type_id = T::fory_get_type_id(context.get_type_resolver())?;
    let remote_type_id = context.reader.read_varuint32()?;
    ensure!(
        local_type_id == remote_type_id,
        Error::type_mismatch(local_type_id, remote_type_id)
    );
    let is_named_enum = local_type_id & 0xff == TypeId::NAMED_ENUM as u32;
    if !is_named_enum {
        return Ok(());
    }
    if context.is_share_meta() {
        // Read type meta inline using streaming protocol
        let _type_info = context.read_type_meta()?;
    } else {
        let _namespace_msb = context.read_meta_string()?;
        let _type_name_msb = context.read_meta_string()?;
    }
    Ok(())
}

pub trait NamedEnumVariantMetaTrait: 'static {
    fn fory_get_sorted_field_names() -> &'static [&'static str] {
        &[]
    }

    #[allow(unused_variables)]
    fn fory_fields_info(type_resolver: &TypeResolver) -> Result<Vec<FieldInfo>, Error> {
        Ok(Vec::default())
    }
}

File: go/fory/enum.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fory

import (
	"reflect"
)

// enumSerializer serializes Go enum types (which are typically int-based types)
// For xlang serialization, enums are written as Varuint32Small7 of their ordinal value
type enumSerializer struct {
	type_  reflect.Type
	typeID uint32 // Full type ID including user ID
}

func (s *enumSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
	// Convert the enum value to its integer ordinal
	var ordinal uint32
	switch value.Kind() {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		ordinal = uint32(value.Int())
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		ordinal = uint32(value.Uint())
	default:
		ctx.SetError(SerializationErrorf("enum serializer: unsupported kind %v", value.Kind()))
		return
	}
	ctx.buffer.WriteVaruint32Small7(ordinal)
}

func (s *enumSerializer) Write(ctx *WriteContext, refMode RefMode, writeType bool, hasGenerics bool, value reflect.Value) {
	if refMode != RefModeNone {
		ctx.buffer.WriteInt8(NotNullValueFlag)
	}
	if writeType {
		// For NAMED_ENUM, need to write type info including namespace and typename meta strings
		typeInfo, err := ctx.TypeResolver().getTypeInfo(value, true)
		if err != nil {
			ctx.SetError(FromError(err))
			return
		}
		ctx.TypeResolver().WriteTypeInfo(ctx.buffer, typeInfo, ctx.Err())
	}
	s.WriteData(ctx, value)
}

func (s *enumSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
	err := ctx.Err()
	ordinal := ctx.buffer.ReadVaruint32Small7(err)
	if ctx.HasError() {
		return
	}

	// Set the value based on the underlying kind
	switch value.Kind() {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		value.SetInt(int64(ordinal))
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		value.SetUint(uint64(ordinal))
	default:
		ctx.SetError(DeserializationErrorf("enum serializer: unsupported kind %v", value.Kind()))
	}
}

func (s *enumSerializer) Read(ctx *ReadContext, refMode RefMode, readType bool, hasGenerics bool, value reflect.Value) {
	err := ctx.Err()
	if refMode != RefModeNone {
		if ctx.buffer.ReadInt8(err) == NullFlag {
			return
		}
	}
	if readType {
		_ = ctx.buffer.ReadVaruint32Small7(err)
	}
	if ctx.HasError() {
		return
	}
	s.ReadData(ctx, value)
}

func (s *enumSerializer) ReadWithTypeInfo(ctx *ReadContext, refMode RefMode, typeInfo *TypeInfo, value reflect.Value) {
	s.Read(ctx, refMode, false, false, value)
}

func (s *enumSerializer) GetType() reflect.Type {
	return s.type_
}
